Semantic model tuple types

chsienki opened the pull request:
Correctly report nullability of tuples in the semantic model: 
- Store the original tuple literal rather than just its type
- Visit the original literal without diagnostics only in the nullability walker to understand its type
- Update the generator to allow force visiting fields in the nullable re-writer
- Update various tuple tests to be correct

Related: #35010

chsienki commented:
/cc @gafter as I understand that you've been working on tuple equality updates. This PR changes the BoundConvertedTupleLiteral slightly; hopefully it doesn't step on your toes too much?

chsienki commented:
@dotnet/roslyn-compiler for review please

333fred commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -1527,8 +1527,8 @@
-    <!-- Natural type is preserved for the purpose of semantic model. Can be null -->
-    <Field Name="NaturalTypeOpt" Type="TypeSymbol" Null="allow"/>
+    <!-- Original tuple is preserved for the purpose of semantic model -->
+    <Field Name="SourceTuple" Type="BoundTupleLiteral" SkipInVisitor="true" ForceInNullabilityRewriter="true" />
```
>ForceInNullabilityRewriter [](start = 76, length = 26)

Perhaps `UpdateFieldInNullabilityRewriter` instead, so it doesn't sound like it conflicts with `SkipInNullabilityRewriter`

333fred replied:
Or perhaps we can update `SkipInVisitor` to accept more than just true/false

---
In reply to: [296056473](https://github.com/dotnet/roslyn/pull/36628#discussion_r296056473) [](ancestors = 296056473)

333fred commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -3986,6 +3986,13 @@ public override BoundNode VisitTupleLiteral(BoundTupleLiteral node)
+            // Disable diagnostics, as we don't want to duplicate any that are produced by visiting the actual literal below
+            var previousDisableDiagnostics = _disableDiagnostics;
+            _disableDiagnostics = true;
+            VisitTupleExpression(node.SourceTuple);
```
Can we use `VisitWithoutDiagnostics` instead?

333fred commented:
Can you please add some tests of the public API to show that this is correctly showing the differences between `Nullability` and `ConvertedNullability`?

333fred commented:
Done review pass (commit 2)

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -52154,14 +52154,12 @@ void M()
     static void F()
     {
-        (object? a, string) t = (default, default)/*T:<null>!*/; // 1
+        (object? a, string) t = (default(object), default(string)) /*T:(object?, string?)*/; // 1
```
>(default(object), default(string)) [](start = 32, length = 34)

Is it necessary to change the test?

chsienki replied:
Yes. `(default, default)` doesn't actually have a type, it's target typed from the LHS. Changing the tests allows us to check that we're correctly returning the type we expect to.

---
In reply to: [296373455](https://github.com/dotnet/roslyn/pull/36628#discussion_r296373455) [](ancestors = 296373455)

chsienki replied:
Will add to the test where we use target typing too

---
In reply to: [296394150](https://github.com/dotnet/roslyn/pull/36628#discussion_r296394150) [](ancestors = 296394150,296373455)

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -52198,30 +52196,28 @@ static void F()
         t.z.ToString();
         if (x == null) return;
-        (object x, object y) u = (x, default)/*T:<null>!*/; // 4
+        (object x, object y) u = (x, default(object))/*T:(object! x, object?)*/; // 4
```
>default(object) [](start = 37, length = 15)

Is it necessary to change the test?

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -52543,8 +52528,8 @@ static void F(bool b, object x)
                 // (6,13): warning CS8619: Nullability of reference types in value of type '(object, object, object, object?, object?, (object, object?), object, object, object?, object?)' doesn't match target type '(object?, object, object?, object, object?, (object, object?), object, object, object?, object)'.
-                //             (x, x, x, default, default, default, x, x, default, default)/*T:<null>!*/; // 1
-                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "(x, x, x, default, default, default, x, x, default, default)").WithArguments("(object, object, object, object?, object?, (object, object?), object, object, object?, object?)", "(object?, object, object?, object, object?, (object, object?), object, object, object?, object)").WithLocation(6, 13),
+                //             (x, x, x, default(object), default(object), default((object, object?)), x, x, default(object), default(object))/*T:(object!, object!, object!, object?, object?, (object!, object?), object!, object!, object?, object?)*/; // 1
```
>default(object) [](start = 41, length = 15)

Same question.

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -53105,6 +53090,66 @@ static void G(object? x, object? y)
+    static void F(string? x)
+    {
+     //   C/*T:C!*/ a = x/*T:string?*/;
+        //(C, C) a = (new C(), new C());
```
Should these lines be uncommented or removed?

chsienki replied:
Mistakenly included, will remove, thanks :)

---
In reply to: [296374978](https://github.com/dotnet/roslyn/pull/36628#discussion_r296374978) [](ancestors = 296374978)

chsienki commented:
Added NullablePublicAPITests.TupleAssignment and updated the type checker to allow checking for converted type too.

---
In reply to: [504242517](https://github.com/dotnet/roslyn/pull/36628#issuecomment-504242517) [](ancestors = 504242517)

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -105312,7 +105438,7 @@ class Program
         }
 
-        // C<T~> operator op(C<T!>, C<T~>) where T : class
+        // C<T~> operator op(C<T!>, C<T~>) where T : class 2
```
> 2 [](start = 58, length = 2)

Is this change intentional?

chsienki replied:
Hah. No. Oh man am I ready for the weekend apparently. Thanks

gafter commented:
I am changing the shape of `BoundConvertedTupleLiteral` in my pending work.  I am not sure how much of a conflict this will cause.

---
In reply to: [504223575](https://github.com/dotnet/roslyn/pull/36628#issuecomment-504223575) [](ancestors = 504223575)

gafter commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -1527,8 +1527,8 @@
-    <!-- Natural type is preserved for the purpose of semantic model. Can be null -->
-    <Field Name="NaturalTypeOpt" Type="TypeSymbol" Null="allow"/>
+    <!-- Original tuple is preserved for the purpose of semantic model -->
+    <Field Name="SourceTuple" Type="BoundTupleLiteral" SkipInVisitor="ExceptNullabilityRewriter" />
```
>SourceTuple [](start = 17, length = 11)

I will have to make this field optional when merging, as nodes of `BoundConvertedTupleLiteral` need to be created when lowering tuple conversions and tuple equality.  We currently generate `BoundTupleLiteral`, but that isn't correct as such nodes should be gone by the end of initial binding and we don't want to have to deal with both forms.
